kernel/fs/iso9660.inc

```diff
@@ -736,3 +736,1044 @@ cd_compare_name:
+        ; create new offset
+        add     dword[esp + 1*4], not -2048
+        and     dword[esp + 1*4], -2048
+
+        cmp     dword[esp + 2*4], edx  ; copy data > read in this sector
+        jae     @f
+        mov     edx, [esp + 2*4]
+@@:
+        sub     dword[esp + 2*4], edx
+        add     dword[esp], edx
+
+        ;DEBUGF  1, "K : iso c=%x d=%x Hz\n", ecx, edx
+        push    esi
+        ; copy (2048 - offset) and -2048
+        mov     esi, ebx ; buffer
+        add     esi, ecx
+        mov     ecx, edx
+        rep movsb
+        pop     esi
+
+        ;stdcall kernel_free, ebx
+.first_align:
+
+        mov     ecx, [esp + 2*4]
+        and     ecx, -2048
+
+        cmp     ecx, 2048
+        jb      .copy_finish_block
+
+        mov     eax, [esp + 1*4]
+        shr     eax, BSF 2048
+        ; copy main block
+        mov     ebx, edi
+        add     edi, ecx
+        sub     dword[esp + 2*4], ecx
+        add     dword[esp + 1*4], ecx
+        add     dword[esp], ecx
+
+        shr     ecx, BSF 2048
+        xor     edx, edx
+        add     eax, esi ; ISO9660_DIRECTORY_RECORD.lba
+        ; ebx - buffer
+        ; edx:eax - num sector
+        ; ebp - PARTITION
+        ; ecx - number sectors
+        call    fs_read64_app
+        test    eax, eax
+        jnz     .err_disk
+
+.copy_finish_block:
+
+        cmp     dword[esp + 2*4], 0
+        jz      .end_align ; creat buffer for end read sector
+
+        mov     ebx, [esp + 4*4]
+
+        ;copy finish block
+        mov     eax, [esp + 1*4]
+        shr     eax, BSF 2048
+        xor     edx, edx
+
+        mov     ecx, 1
+        add     eax, esi ; ISO9660_DIRECTORY_RECORD.lba
+        ; ebx - buffer
+        ; edx:eax - num sector
+        ; ebp - PARTITION
+        ; ecx - number sectors
+        call    fs_read64_app
+        test    eax, eax
+        jnz     .err_disk_1
+
+        mov     esi, ebx
+        mov     ecx, [esp + 2*4]
+        add     dword[esp], ecx
+        rep movsb
+
+        ;stdcall kernel_free, ebx
+
+.end_align:
+        ; set ebx size copy data
+        mov     ebx, [esp]
+        mov     esi, [esp + 3*4]
+        add     esp, 4*4
+        call    kernel_free
+
+        mov     eax, esi
+        ret
+
+.err_disk_1:
+        stdcall kernel_free, ebx
+.err_disk:
+        add     esp, 4*4
+        call    kernel_free
+        xor     ebx, ebx
+        mov     eax, EIO
+        ret
+
+;-----------------------------------------------------------------------------
+; ISO9660 external functions
+;   in:
+; ebx -> parameter structure of sysfunc 70
+; ebp -> ISO9660 structure
+; esi -> path string in UTF-8
+;   out:
+; eax, ebx = return values for sysfunc 70
+iso9660_ReadFolder:
+        sub     esp, 4 ; for ptr on memory page
+        call    iso9660_find_file
+
+        test    byte[eax + ISO9660_DIRECTORY_RECORD.flags], 10b ; check dir
+        jz      iso9660_find_file.not_found
+
+        mov     edi, [ebx + 16] ; buffer
```

Comment:
Replace with structure fields?

Reply:
there is no clear definition of the structure for working with files in the kernel, which is why magic numbers are used